src/mlpack/methods/matrix_completion/matrix_completion_impl.hpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * @file matrix_completion_impl.hpp
+ * @author Stephen Tu
+ * @author Chenzhe Diao
+ *
+ * Implementation of MatrixCompletion class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_MATRIX_COMPLETION_MATRIX_COMPLETION_IMPL_HPP
+#define MLPACK_METHODS_MATRIX_COMPLETION_MATRIX_COMPLETION_IMPL_HPP
+
+#include "matrix_completion.hpp"
+
+namespace mlpack {
+namespace matrix_completion {
+
+template<typename MCSolverType>
+MatrixCompletion<MCSolverType>::
+MatrixCompletion(const size_t m,
+                 const size_t n,
+                 const arma::umat& indices,
+                 const arma::vec& values,
+                 const size_t r) :
+    m(m), n(n), mcSolver(m, n, indices, values, r)
+{
+  CheckValues(indices, values);
+}
+
+template<typename MCSolverType>
+MatrixCompletion<MCSolverType>::
+MatrixCompletion(const size_t m,
+                 const size_t n,
+                 const arma::umat& indices,
+                 const arma::vec& values,
+                 const arma::mat& initialPoint) :
+    m(m), n(n), mcSolver(m, n, indices, values, initialPoint)
+{
+  CheckValues(indices, values);
+}
+
+template<typename MCSolverType>
+MatrixCompletion<MCSolverType>::
+MatrixCompletion(const size_t m,
+                 const size_t n,
+                 const arma::umat& indices,
+                 const arma::vec& values) :
+    m(m), n(n), mcSolver(m, n, indices, values)
+{
+  CheckValues(indices, values);
+}
+
+template<typename MCSolverType>
+MatrixCompletion<MCSolverType>::
+MatrixCompletion(const size_t m,
+                 const size_t n,
+                 const arma::umat& indices,
+                 const arma::vec& values,
+                 const double tau) :
+    m(m), n(n), mcSolver(m, n, indices, values, tau)
+{
+  CheckValues(indices, values);
+}
+
+
+template<typename MCSolverType>
+void MatrixCompletion<MCSolverType>::
+CheckValues(const arma::umat& indices, const arma::vec& values)
```

Comment:
It might be possible to simply rely on Armadillo throwing an exception in the `sp_mat` constructor instead of having this `CheckValues()` function; I know that function will fail if the given matrix is of the wrong size.